Feature: introduce "Window Mode" for selecting windows under Hyprland

Syndelis opened the pull request:
## About the PR

Resolves #4 

This PR introduces two new feature flags: `window-selection` and `hyprland-window-selection`. The former provides an API for compositor-specific implementations, while the latter is an implementation for Hyprland.

The `window-selection` feature allows for a new mode to be used, "Window Mode", which no longer allows for the mouse to draw selections, but rather to have the selection be drawn around clicked windows (much similar to how "Monitor Mode" works).

<kbd>Tab</kbd>'s cycle order for mode is now Display → Window → Rectangle. Additionally, going from "Window Mode" to "Rectangle Mode" preserves the selection around the last selected window. It's worth noting that this is only true when the flag is enabled, otherwise all remains the same.

In addition to this mode, new command line options are introduced to allow for a non-interactive usage of the program. Although it was noted on #4 that one should use `grim` directly when interactiveness is not desired, now that the program is able to find windows I think this becomes relevant again. The command line options are mutually exclusive and listed below:
- `--active-window`: Pre-selects the currently focused window;
- `--window-under-cursor`: Pre-selects the window that's under the cursor. This may be different than focused window depending on the user's compositor and settings, but for Hyprland they yield the same results;
- `--window-search`: Pre-selects the first window that matches a condition. The condition is made up of a search parameter and a search value. The search parameter can be the window's title, class, initial title or initial class. The search value is a regex which will be matched against the search paramenter;

While these 3 mutually exclusive options work fine for interactive work (i.e. the program's default behavior), they can be used along with the new CLI argument `--auto-capture` to allow for non-interactive window capturing. For example, a simple command to capture the currently focused window can be achieved with `watershot --active-window --auto-capture directory ~/Pictures`.

## Extras

This PR also updates the `wgpu` and `wgpu_text` dependencies. Rustc kept complaining that `wgpu_text::section` and `wgpu_text::font` didn't exist and, while I couldn't find any PR mentioning this change, I could find that the re-exports for what was used in this project were available under `wgpu_text::glyph_brush`.

## Closing Note

I've tried my best to abstract away the Hyprland-specific stuff under generic APIs so that others can implement the same functionalities for Sway, GNOME, etc. without having the re-invent the wheel. I hope that it achieves its goal!

GabrielFoxDev commented:
lgtm!

Kirottu commented on src/runtime_data.rs:
```diff
@@ -58,10 +61,13 @@ pub struct RuntimeData {
 
     pub renderer: Renderer,
+
+    #[cfg(feature = "window-selection")]
```
Could be replaced with a boxed unified backend implementation.

Syndelis replied:
Addressed in c52ba23

Kirottu reviewed (commented):
In general this is neat work, if you want I can address my comments myself or you can also tackle them by yourself if you wish.

Syndelis commented on src/window/mod.rs:
```diff
@@ -0,0 +1,51 @@
+    fn get_mouse_position() -> (i32, i32);
+}
+
+#[cfg(feature = "hyprland-window-selection")]
```
That's a great suggestion! I did experiment with using trait objects at first, but I figured one would only have one implementation in use at any given time. To be honest, the thought of deciding the implementation at runtime didn't even cross my mind, but it makes a lot of sense! I'll work on that ASAP

Syndelis replied:
Addressed in c52ba23

Syndelis commented on src/window/mod.rs:
```diff
@@ -0,0 +1,51 @@
+pub type WindowDescriptor = hyprland::HyprWindowDescriptor;
+
+#[cfg(feature = "hyprland-window-selection")]
+pub type MouseGetter = hyprland::HyprMouseGetter;
```
That would be great as it lowers the barrier for new implementations to be added. Can you help me on that one?

Kirottu replied:
I am not sure if the pointer capability stored in the `RuntimeData` struct reports positions directly, but at least the events for movement are received and used for selection already. It would just be a matter of translating it into the needed coordinate space to find windows using the already present method.

Syndelis replied:
I took this evening to address most suggestions, but this one puzzled me. The pointer in `RuntimeData` seems to be set only in `new_capability`, but that method is never called. Did I get that wrong?

Kirottu replied:
Watershot itself does not call it, that is right. It is a Wayland event received by SCTK in the main event loop, and then called by the same event loop.

Syndelis replied:
As discussed [here](https://github.com/Kirottu/watershot/pull/20#discussion_r1285235475), I've moved the mouse-getting function to the `CompositorBackend` trait in this commit 80d3246

Syndelis commented on src/window/mod.rs:
Sure thing, makes sense to me; especially if we're turning down the mouse trait

Syndelis replied:
Addressed in c2ca665, 80d3246 and ccf3f80

Syndelis commented on src/main.rs:
```diff
@@ -14,6 +14,10 @@ mod macros;
 mod traits;
 mod types;
+
+#[cfg(feature = "window-selection")]
```
Maybe it would still make sense if there is no backend available for the user's setup? We could write a build.rs to determine that. Would love your opinion on that!

Kirottu replied:
People don't really build stuff themselves. They generally use easy methods like the AUR and stuff so having to make multiple packages for people that do have a backend and people who don't doesn't make much sense to me. Especially as the cost of just having the implementations in the binary is negligible.

Syndelis replied:
Addressed in c52ba23

Syndelis commented on src/types.rs:
```diff
@@ -312,17 +341,46 @@ pub enum SelectionModifier {
+}
+
+impl Selection {
+    pub fn flattened(&self) -> Selection {
```
I think the easiest of the options is to convert it into a rectangle selection, so I'll get working on that! We can revisit the handles if it turns out too confusing down the road

Syndelis replied:
Addressed in c53340d

Syndelis commented:
> In general this is neat work, if you want I can address my comments myself or you can also tackle them by yourself if you wish.

Thanks for the review, @Kirottu! While I can handle most comments myself fairly quickly, I'd like to ask you for a general direction about the mouse coordinates you mentioned!

Syndelis commented on src/window/hyprland.rs:
```diff
@@ -0,0 +1,117 @@
+
+        Clients::get()
+            .unwrap()
+            .iter()
```
Addressed in c52ba23

Syndelis commented:
I've made some commits addressing most suggestions, @Kirottu, but did leave one question about the mouse position one.

Also, I didn't want to mark anything resolved because you may want to leave new comments on those ones, so feel free to "Mark resolved" whatever you feel has been correctly addressed!

Kirottu commented on src/runtime_data.rs:
```diff
@@ -111,14 +125,50 @@ impl RuntimeData {
+            (selection, windows, exit) = {
+                let windows = compositor_backend.get_all_windows();
+
+                let selection = match (args.window_search.take(), args.window_under_cursor, args.active_window) {
```
The `group` option in the `clap` macros already makes these mutually exclusive, so this looks pretty messy and would probably be nicer to look at if turned into a simple chain of `if` statements.

Syndelis replied:
Addressed in 0628de3

Kirottu commented on src/types.rs:
```diff
@@ -296,6 +316,10 @@ impl Rect<i32> {
         }
     }
+
+    pub fn contains_point(&self, x: i32, y: i32) -> bool {
```
A `Contains` trait implementation should probably be used instead.

Syndelis replied:
Definitely missed that

Syndelis replied:
Addressed in 45bcc26

Kirottu commented on src/window/mod.rs:
```diff
@@ -0,0 +1,66 @@
+    fn try_new() -> Result<Box<dyn CompositorBackend>, CompositorNotAvailable>;
+}
+
+pub trait DescribesWindow: DescribesWindowClone {
```
I feel like having this as a trait at all isn't that useful, I think the type that describes a window should be compositor agnostic as there isn't exactly any compositor specific functionality there that Watershot would need.

The compositor specific part is obtaining the window info (`CompositorBackend`), not the window info itself.

Syndelis replied:
Addressed in c2ca665

Kirottu commented on src/window/mod.rs:
```diff
@@ -0,0 +1,66 @@
+    }
+}
+
+pub enum CompositorNotAvailable {
```
If Watershot is running at all, a compositor is running as well so a more descriptive error would be something like unsupported, or something custom reported by the backend.

Syndelis replied:
This enum is used as an error type for `InitializeBackend::try_new`. The reasoning is that, if we're not feature-gating compositor implementations, we need to check at runtime **which** compositor is available to be used (not *whether* any compositor is running). The function I mentioned is used in `RuntimeData::get_preferred_backend`, where it currently always returns Hyprland or nothing. With more backend implemenations, I could see it looking like this:

```rust
pub fn get_preferred_backend() -> Option<Box<dyn CompositorBackend>> {
    HyprlandBackend::try_new().or(SwayBackend::try_new()).or(GnomeBackend::try_new()).ok()
}
```

Kirottu replied:
Yes I know how the enum is used, but the `NotRunning` and `NotInstalled` variants don't really feel like they describe real world problems, at least when Watershot is already running.

Syndelis replied:
The problem I believe it describes happens when the user launches Watershot and the program needs to decide whether or not the window selection mode should be available, by asking "Is there a supported compositor backend installed and running?"

For example, if a Sway user launches Watershot, they shouldn't be able to use "Window Mode" because there is no Sway backend implementation as it stands. The program only knows that because it tried to initialize all backend implementations and none could be initialized.

So, *a* compositor is running (Sway) but *the* window-mode-supported compositor (Hyprland) is not.

Kirottu replied:
Ah yeah that does make sense for `NotRunning`.

Kirottu commented on src/sctk_impls/pointer_handler.rs:
```diff
@@ -94,38 +94,46 @@ impl PointerHandler for RuntimeData {
+
+                            if win_sel.is_some() {
+                                self.selection = Selection::Window(win_sel);
+                            } else {
```
I think the handles should be tried first here, with this approach you can easily accidentally click inside the bounds of the window while trying to grab the handle which will make the handle not work. So that if there is a selection already, check the handles, if changed then convert the selection. If not changed or if there was no selection then try to create a new selection.

Syndelis replied:
Addressed in 4657999

Kirottu commented on src/sctk_impls/keyboard_handler.rs:
```diff
@@ -52,34 +52,47 @@ impl KeyboardHandler for RuntimeData {
+                        self.selection = Selection::Rectangle(None)
+                    }
+                }
+                Selection::Window(_) => self.selection = self.selection.flattened(),
```
I think if we are deliberately changing the mode, the selection should be cleared.

Syndelis replied:
Sure thing; since moving the handles now keeps the selecion I guess that functionality is already there

Syndelis replied:
Addressed in e515381

Kirottu commented on src/runtime_data.rs:
```diff
@@ -111,14 +125,50 @@ impl RuntimeData {
+
+                let selection = match (args.window_search.take(), args.window_under_cursor, args.active_window) {
+                    (None, true, false) => {
+                        let (mouse_x, mouse_y) = MouseGetter::get_mouse_position();
```
I also completely forgot about this argument, it does make getting rid of compositor specific mouse coordinates much harder if we want to keep it.

Syndelis replied:
Indeed, that was what kept me from removing it. I'll at least move it to the `CompositorBackend` trait, and we can try removing it in the future

Syndelis replied:
Addressed in 80d3246

Syndelis commented on src/main.rs:
```diff
@@ -1,3 +1,5 @@
+#![feature(return_position_impl_trait_in_trait)]
```
Oh sorry, that's a leftover, I'm not even using that

Kirottu replied:
Yeah I noticed it when I saw that no traits actually had return position `impl`s.

Syndelis replied:
Addressed in d2fffc8

Syndelis commented:
I believe I've managed to address all the review comments this time around, please double check though!

Once again, thanks for all the suggestions, @Kirottu !

Kirottu commented:
There seems to be a bit of a regression with the ability to move a rectangular selection from its center area being broken now.

Syndelis commented on src/window/mod.rs:
```diff
@@ -0,0 +1,54 @@
+    fn try_new() -> Result<Box<dyn CompositorBackend>, CompositorNotAvailable>;
+}
+
+pub trait FindWindowExt {
```
I mainly plan to use it to bind a hotkey to take screenshots of a chrome instance in any page of my company's software. I'm frequently investigating bugs and I tend to take a bunch of screenshots as proof. With the hotkey, I hope I'm gonna hasten this experience

Kirottu replied:
Alright fair's fair.

Syndelis commented:
> There seems to be a bit of a regression with the ability to move a rectangular selection from its center area being broken now.

I'll take a look at it

Kirottu commented:
Alright, I think this is in good enough shape to be merged now. Many thanks for the great work!
